File: 2000 98 & ME/QuickStart/howto/samples/cryptography/fileencrypt/encrypt.cs

// This program uses a passphrase-derived key and DES encryption to encrypt a test file then decrypt it.


using System;
using System.IO;
using System.Security;
using System.Security.Cryptography;

public class Encryption
{
    static private byte[] symkey;
    static private byte[] symIV;
    static private String passPhrase = "YourDog123";
    static private String inputFileName = "input.txt";
    static private String encrypedFileName = "encrypted.txt";
    static private String deencrypedFileName = "decrypted.txt";
    
    public static void Main(string[] args) {
        
        
        Console.WriteLine ("Creating Key");
        //fills in symKey and symIV arrays
        GenerateKeyFromPassPhrase(passPhrase);
        
        Console.WriteLine ("Encrypting {0}", inputFileName);
        EncryptDecryptData(inputFileName, encrypedFileName);
        Console.WriteLine ("Decrypting {0}", encrypedFileName);
        EncryptDecryptData(encrypedFileName, deencrypedFileName);

		Console.WriteLine ("\r\nPress Return to exit.");
		Console.Read();
    }
    
    ///////////////////////////////////////////////////////////////
    // Compute DES Key and IV Values
    static private bool GenerateKeyFromPassPhrase(String phrase)
    {
        try {
            
            // change the pass phrase into a form we can use
            int i;
            int len;
            char[] cp = phrase.ToCharArray();
            len = cp.GetLength(0);
            // convert to byte array -  discard upper byte since is usually 0 in Eng.
            byte[] bp = new byte[len];
            for (i=0; i<len; i++) bp[i] = (byte)cp[i]; // truncate char value
            
            // Crypto specific code starts here
            // allocate the key and IV arrays, assumes DES size
            symkey = new byte[8]; 
            symIV = new byte[8];
            //do the hash operation
            SHA1_CSP sha = new SHA1_CSP();
            sha.Write(bp);
            sha.CloseStream();
            // Crypto specific code ends here
            // use the low 64-bits for the key value
            for (i=0; i<8; i++) symkey[i] = sha.Hash[i];
            for (i=8; i<16; i++) symIV[i-8] = sha.Hash[i];
            
            return true;
        }
        catch (Exception e) {
            Console.WriteLine(e);
            return false;
        }
    }
    
    ///////////////////////////////////////////////////////////////
    // Encrypt/Decrypt Data
    static private void EncryptDecryptData(String inName, String outName)
    {
        try {
            bool bEncrypt = true;
            
            // Open the input and output files
            FileStream fin = new FileStream(inName, FileMode.Open, FileAccess.Read);
            
            FileStream fout = new FileStream(outName, FileMode.OpenOrCreate, FileAccess.Write);
            fout.SetLength(0);
            
            //setup up variable to help with read
            byte[] bin = new byte[4096];
            long totlen = fin.Length; 
            long rdlen = 8;
            int len;
            int i;
            
            // figure out if we're encrypting or decrypting
            byte[] tag = new byte[8];
            fin.Read(tag,0,8);
            if ((tag[0]==(byte)'[') && (tag[1]==(byte)'E') && (tag[2]==(byte)'n') && (tag[3]==(byte)'c') && (tag[4]==(byte)']')) {
                Console.WriteLine("Decrypting data");
                bEncrypt = false;
            }
            else {
                Console.WriteLine("Encrypting");
                fin.Seek(0, SeekOrigin.Begin);
                rdlen = 0;
            }
            
            // Crypto specific code starts here
            // create the DES encryptor and the SymmetricStreamEncryptor obj
            DES_CSP des = new DES_CSP();
            StoreCryptoStream scs = new StoreCryptoStream(bEncrypt,fout);
            
            if (bEncrypt) {
                SymmetricStreamEncryptor sse = des.CreateEncryptor(symkey, symIV);
                
                // a little extra feature here to show how to compose crypto 
                // components that support ICryptoStream
                SHA1_CSP sha = new SHA1_CSP();
                
                // wire up the encryptor - hash - StoreCryptoStream
                sse.SetSink(sha);	
                sha.SetSource(sse);
                sha.SetSink(scs);	
                scs.SetSource(sha);
                while (rdlen < totlen) {
                    len = fin.Read(bin,0,4096);
                    sse.Write(bin,0,len);
                    rdlen = rdlen + len;
                    Console.WriteLine("Processed {0} bytes, {1} bytes total",len, rdlen);
                }
                sse.CloseStream();
                Console.WriteLine("HASH of the encrypted stream is:");
                Console.Write("\t");
                for (i=0; i< 20; i++) Console.Write("{0} ",sha.Hash[i]);
				Console.WriteLine();
            }
            else {
                SymmetricStreamDecryptor ssd = des.CreateDecryptor(symkey, symIV);
                ssd.SetSink(scs);
                scs.SetSource(ssd);
                while (rdlen < totlen) {
                    len = fin.Read(bin,0,4096);
                    ssd.Write(bin,0,len);
                    rdlen = rdlen + len;
                    Console.WriteLine("Processed {0} bytes, {1} bytes total",len, rdlen);
                }
                ssd.CloseStream();
            }
            // Crypto specific code ends here
            
            fin.Close();
        }
        catch (Exception e) {
            Console.WriteLine(e);
            return;
        }
    }
}

/*******************************************************************
* Helper class for capturing the output from the symmetric encryptor
* or decryptor and writing it out to the target file
*******************************************************************/
public class StoreCryptoStream : ICryptoStream
{
    
    static byte[] tag = {(byte)'[',(byte)'E',(byte)'n',(byte)'c',(byte)']',
        0x20, 0x20, 0x20};
    FileStream fs;
    
    ////////////////////////////////////////////////////////////////
    // Constructor
    public StoreCryptoStream(bool bEncrypt, FileStream fout) 
    {
        fs = fout;
        if (bEncrypt) fs.Write(tag,0,8);
    }
    
    public virtual void CloseStream() {fs.Close();}
    public virtual void CloseStream(Object obj) {fs.Close();}
    public virtual void SetSink(ICryptoStream pstm) {}
    public virtual void  SetSource(CryptographicObject co) {}
    public virtual ICryptoStream  GetSink () {return null;}
    
    ////////////////////////////////////////////////////////////////
    // Write routines just copy output to the target file
    public virtual void Write(byte[] bin)
    {
        int len = bin.GetLength(0);
        Write(bin, 0, len);
    }
    
    public virtual void Write(byte[] bin, int start, int len )
    {
        fs.Write(bin,start,len);
    }

}

File: 2000 98 & ME/QuickStart/howto/samples/cryptography/fileencrypt/VB/encrypt.vb
' This program uses a passphrase-derived key and DES encryption to encrypt a test file then decrypt it.

Imports System
Imports System.IO
Imports System.Security
Imports System.Security.Cryptography
Imports Microsoft.VisualBasic

Namespace ClassLibrary1
    Public Class encrypt1
        Private Shared symkey(8) As Byte
        Private Shared symIV(8) As Byte
        Private Shared passPhrase As String = "YourDog123"
        Private Shared inputFileName As String = "input.txt"
        Private Shared encrypedFileName As String = "encrypted.txt"
        Private Shared deencrypedFileName As String = "decrypted.txt"
    
        Public Shared Sub Main()
            Console.WriteLine("Creating Key")
            'fills in symKey and symIV arrays
            GenerateKeyFromPassPhrase(passPhrase)
        
            Console.WriteLine("Encrypting {0}", inputFileName)
            EncryptDecryptData(inputFileName, encrypedFileName)
            Console.WriteLine("Decrypting {0}", encrypedFileName)
            EncryptDecryptData(encrypedFileName, deencrypedFileName)

            Console.WriteLine
            Console.WriteLine("Press Return to exit.")
            Console.Read()
        End Sub

        '/////////////////////////////////////////////////////////////
        ' Compute DES Key and IV Values
        Private Shared Function GenerateKeyFromPassPhrase(ByVal phrase As String) As Boolean
            Try
            ' change the pass phrase into a form we can use
                Dim i As Integer
                Dim len As Integer
                Dim cp() As Char = phrase.ToCharArray()
                len = cp.GetLength(0)
                ' convert to byte array -  discard upper byte since is usually 0 in Eng.
                Dim bp(len) As Byte
                For i = 0 To len - 1
                     bp(i) = CByte (cp(i)) ' truncate char value
                Next i

                ' Crypto specific code starts here
                ' allocate the key and IV arrays, assumes DES size
                'Dim symkey(8) As Byte
                'Dim symIV(8) As Byte
                'do the hash operation 
                Dim sha As New SHA1_CSP
                sha.Write(bp)
                sha.CloseStream()
                ' Crypto specific code ends here 
                ' use the low 64-bits for the key value 
                For i = 0 To 7
                    symkey(i) = sha.Hash(i) 
                Next i

                For i = 8 To 15
                    symIV(i - 8) = sha.Hash(i)
                Next i
            
                GenerateKeyFromPassPhrase = True
            Catch e As Exception
                Console.WriteLine(e)
                GenerateKeyFromPassPhrase = False
            End Try
        End Function

        '/////////////////////////////////////////////////////////////
        ' Encrypt/Decrypt Data
        Private Shared Sub EncryptDecryptData(ByVal inName As String, ByVal outName As String)
            Try
				Dim i as Integer
                Dim bEncrypt As Boolean = True
            
                ' Open the input and output files
                Dim fin As FileStream = New FileStream(inName, FileMode.Open, FileAccess.Read)
            
                Dim fout As FileStream = New FileStream(outName, FileMode.OpenOrCreate, FileAccess.Write)
                fout.SetLength(0)
            
                ' setup up variable to help with read
                Dim bin(4096) As Byte
                Dim totlen As Long = fin.Length
                Dim rdlen As Long = 8
                Dim len As Integer

            
                ' figure out if we're encrypting or decrypting 
                Dim tag(8) As Byte
                fin.Read(tag, 0, 8)
                If ((tag(0) = CByte("["c)) And (tag(1) = CByte("E"c)) And (tag(2) = CByte("n"c)) And (tag(3) = CByte("c"c)) And (tag(4) = CByte("]"c))) Then
                    Console.WriteLine("Decrypting data")
                    bEncrypt = False
                Else
                    Console.WriteLine("Encrypting")
                    fin.Seek(0, SeekOrigin.Begin)
                    rdlen = 0
                End If
            
                ' Crypto specific code starts here 
                ' create the DES encryptor and the SymmetricStreamEncryptor obj 
                Dim des As DES_CSP = New DES_CSP
                Dim scs As StoreCryptoStream = New StoreCryptoStream(bEncrypt, fout)
            
                If bEncrypt Then
                    Dim sse As SymmetricStreamEncryptor = des.CreateEncryptor(symkey, symIV)
                
                    ' a little extra feature here to show how to compose crypto  
                    ' components that support ICryptoStream 
                    Dim sha As SHA1_CSP = New SHA1_CSP
                
                    ' wire up the encryptor - hash - StoreCryptoStream 
                    sse.SetSink(sha)
                    sha.SetSource(sse)
                    sha.SetSink(scs)
                    scs.SetSource(sha)
                    While (rdlen < totlen)
                        len = fin.Read(bin, 0, 4096)
                        sse.Write(bin, 0, len)
                        rdlen = rdlen + len
                        Console.WriteLine("Processed {0} bytes, {1} bytes total", len, rdlen)
                    End While
                    sse.CloseStream()
                    Console.WriteLine("HASH of the encrypted stream is:")
					Console.Write(Strings.chr(9))
                    For i = 0 To 19
                        Console.Write("{0} ", sha.Hash(i))
                    Next i
					Console.WriteLine()
                Else
                    Dim ssd As SymmetricStreamDecryptor = des.CreateDecryptor(symkey, symIV)
                    ssd.SetSink(scs)
                    scs.SetSource(ssd)
                    While (rdlen < totlen)
                        len = fin.Read(bin, 0, 4096)
                        ssd.Write(bin, 0, len)
                        rdlen = rdlen + len
                        Console.WriteLine("Processed {0} bytes, {1} bytes total", len, rdlen)
                    End While
                    ssd.CloseStream()
                End If
                ' Crypto specific code ends here 
            
                fin.Close()
        
            Catch e As Exception
                Console.WriteLine(e)
            End Try
        End Sub
    End Class

    Public Class StoreCryptoStream
        Implements ICryptoStream

        Shared tag() As Byte = { CByte("["c), CByte("E"c), CByte("n"c), CByte("c"c), CByte("]"c), &H20, &H20, &H20 }
        Dim fs As FileStream

        '////////////////////////////////////////////////////////////////
        '// Constructor
        Public Sub new(ByVal bEncrypt As Boolean, ByVal fout As FileStream)
            MyBase.new()
            fs = fout
            If bEncrypt Then
                fs.Write(tag, 0, 8)
            End If
        End Sub

        Overloads Public Sub CloseStream() Implements ICryptoStream.CloseStream
            fs.Close()
        End Sub

        Overloads Public Sub CloseStream(ByVal obj As Object) Implements ICryptoStream.CloseStream
            fs.Close()
        End Sub

        Public Sub SetSink(ByVal pstm As ICryptoStream) Implements ICryptoStream.SetSink
        End Sub

        Public Sub SetSource(ByVal co As CryptographicObject) Implements ICryptoStream.SetSource
        End Sub
        
        Public Function  GetSink () As ICryptoStream  Implements ICryptoStream.GetSink
          return Nothing
        End Function

        '////////////////////////////////////////////////////////////////
        '// Write routines just copy output ot the target file
        Overloads Public Sub Write(ByVal bin() As Byte) Implements ICryptoStream.Write
            Dim len As Integer = bin.GetLength(0)
            Write(bin, 0, len)
        End Sub
 
        Overloads Public Sub write(ByVal bin() As Byte, ByVal start As Integer, ByVal len As Integer) Implements ICryptoStream.Write
            fs.Write(bin, start, len)
        End Sub

    End Class


End Namespace